src/Admin/AbstractAdmin.php

```diff
@@ -1734,20 +1748,36 @@ public function setSubject($subject)
 
     public function getSubject()
     {
-        if (null === $this->subject && $this->request && !$this->hasParentFieldDescription()) {
-            $id = $this->request->get($this->getIdParameter());
+        if (!$this->hasSubject()) {
+            @trigger_error(sprintf(
+                'Calling %s() when there is no subject is deprecated since sonata-project/admin-bundle 3.x and will throw an exception in 4.0. '.
+                'Use %s::hasSubject() to know if there is a subject.',
+                __METHOD__,
+                __CLASS__
+            ), E_USER_DEPRECATED);
+            // NEXT_MAJOR : remove the previous `trigger_error()` call, the `return null` statement, uncomment the following exception and update the return type
+            // throw new \LogicException(sprintf(
+            //    'Admin "%s" has no subject.',
+            //    static::class
+            // ));
 
-            if (null !== $id) {
-                $this->subject = $this->getObject($id);
-            }
+            return null;
         }
 
         return $this->subject;
     }
 
     public function hasSubject()
     {
-        return (bool) $this->getSubject();
+        if (null === $this->subject && $this->hasRequest() && !$this->hasParentFieldDescription()) {
+            $id = $this->request->get($this->getIdParameter());
+
+            if (null !== $id) {
+                $this->subject = $this->getObject($id);
```

Comment:
To me I wouldn't expect setting something in a `has` method.

Reply:
Previously it was returning `return (bool) $this->getSubject();`.
So the value was already set by the `has` method. I changed nothing.

Plus I wouldn't expect having `hasSubject` returning `false` and `getSubject` returning something.

Reply:
I'm not saying that before was ok.

Just that IMHO this code should not be in a `hasser`, if it's a temporary solution I guess should be documented how to handle this in the next major version.

Reply:
In next major, a lot of property will be private. IMHO `$subject` will be too.
I see `$this->subject` like a cache. If no value was set, there is a default value coming from the URL.

In the same way, `public function hasListFieldDescription($name)` is calling `buildList`.

SoI don't see this as a temporary solution. It's debatable, of course. 
But since 
- I made no change about the current behaviour
- There is no proof that this should be (or not) a temporary solution
- There is currently no other solution provided
- We don't know how big is the work to change this behaviour 

I think the debate should not be related to this PR and should have his own issue/PR.